### ISSUE TITLE
Static extensions declare constructors with a two-part name, do we want to use a less verbose form?

### ISSUE THREAD

**Issue by eernstg** [dart-lang/language#4046]
This issue raises the question whether we should declare a constructor for a class `C` in an extension using the name `C` or `C.named`, or we should use a more concise syntax.

Background: https://github.com/dart-lang/language/pull/3835 provides a proposal for generalizing `extension` declarations such that they can contribute static members and constructors to an existing declaration (e.g., a class, an extension type, etc.), in a way that may resemble extension instance members (that is, "normal" members of the extension declarations that we have had in the language for several years). The same PR also provides a proposal where a new kind of declaration, `static extension`, delivers the same affordances, but avoids conflating the semantics of `extension` with new mechanisms and new syntax.

We may prefer one or the other proposal, or something else altogether, but the following question seems relevant to them all (and I'll just use `static extension` to refer to the kind of declaration that provides this feature):

A `static extension` can declare a constructor for its so-called _on-class_ using the full name which is also used at call sites (aka instance creation expressions). For example:

```dart
class MyClass<X extends Iterable<Y>, Y> {
  final X x;
  MyClass.namedInC(this.x);
}

static extension E<Z> on MyClass<List<Z>, Z> {
  factory MyClass.namedInE() => C(const []);
  const MyClass.named2InE(Iterable<Z> iter): this.namedInC(iter);
  factory MyClass(Z z) {
    // ...
    return MyClass.namedInC(<Z>[z]);
  }
}
```

We could choose to declare the first factory using `factory namedInE() => ...;`. Even the redirecting generative constructor could be declared using `named2InE(Iterable<Z> iter): this.namedInC(iter);`. A constructor whose name is just `MyClass` (sometimes known as a "nameless" constructor) could be declared using `new(...) ...`. This yields the following result:

```dart
static extension E<Z> on C<List<Z>, Z> {
  factory namedInE() => C(const []);
  const named2InE(Iterable<Z> iter): this.namedInC(iter);
  factory new(Z z) {
    return MyClass.namedInC(<Z>[z]);
  }
}
```

This would yield a syntax which is more concise than the syntax which is in the current proposals (where constructor declaration names always include the class name). The more concise syntax is still unambiguous because we have `factory` in some constructor declarations (which can't be a non-constructor), and we have a colon plus `this(...)` in others (which also can't be a non-constructor).

So @dart-lang/language-team,  what do you prefer in the case of constructor declarations in extensions, brevity or explicitness?

**Comment by lrhn** [comment]
Is there any reason for allowing shorter constructor names in static extensions, and not also allow the same syntax everywhere else?
Likely not, which means that this is not about static extensions, it's just about shorter constructor declarations in general (like #3788).

I prefer to get static extensions without any shorthand over delaying static extensions to do proper constructor shorthands. And until we have decided what "proper constructor shorthands" are, I can't say whether I want them or not. (But I probably do.)

Omitting the class name is slightly more *relevant* to static extension, where it's not absolutely clear whether to write `OnClass.name(...)` or `StaticExtensionName.name(...)` for a constructor. We can allow both. We can require either. You can most likley *call* the constructor using either naming, with `StaticExtensionName.name(...)` being a fallback access in case of naming conflicts.
Just omitting the class name makes that choice irrelevant if you write *neither* name.

Anyway, options include:
* Use `new` instead of class name: `new();`, `new.name()`, `factory new()`, `factory new.name()`, `const new()` *cough*. Oh well.
* Use `new` or `const` instead of class name: `new()`, `new.name()`, `factory new()`, `const()`, `const.name()`, `factory const()` ... puts `const` after `factory`, unlike today, which can be confusing.
* Same, but use `factory` instead of `new` for non-const factory constructors: `new()`,.., `factory()`, ... `const factory()` *or* `factory const()`. Ick, probably not both.
* Use space instead of dot for names: `new name()`, `const name()`. No, I want to reserve `const foo()` for constant functions. Constructors are not the only things that *could* be constant.

(I think "`new` or `const` instead of class name" is the best trade-off. And it obviously wouldn't work for primary constructors, where the same declaration declares the class name and constructor base name.)

TL;DR: I prefer static extension ASAP over any syntactic shorthands. And any shorthand should apply to *all* constructors, not just static extensions.

**Comment by eernstg** [comment]
> Is there any reason for allowing shorter constructor names in static extensions, and not also allow the same syntax everywhere else?

Good point, no there is no reason. The argument was raised for static extensions, and I wanted to make sure the topic wasn't thrown away without due consideration, so I created this issue.

But now I'll close the issue as a duplicate of https://github.com/dart-lang/language/issues/3788. I don't see any reason to do anything special about the naming of constructors in static extensions.
